### DISCUSSION TITLE
How do I reuse cache properly? for OpenCode agent switching

### DISCUSSION THREAD

**Discussion by Nindaleth** [ggml-org/llama.cpp#22354] [Q&A] [answered]
My use case is working with OpenCode - I create a plan with a Plan agent and then switch to Build agent and "please implement the outlined plan" - each agent has a different system prompt which thus triggers a full prompt reprocessing, despite more than a half of the cache (user prompt, assistant messages and tool call outputs) not changing at all.

I've noticed there's a `--cache-reuse` which should be exactly what I need, but it doesn't seem to work. Here's my server command that I use for testing - it's one already proven to work nicely, now with greedy decoding, fixed seed and  `--cache-reuse 256` appended:

```
./build/bin/llama-server --host 0.0.0.0 --port 1234 --api-key-file api_keys.txt --sleep-idle-seconds 600 \
--jinja --flash-attn on --parallel 2 --kv-unified --checkpoint-every-n-tokens 13000 --ctx-checkpoints 128 \
--cache-ram 16384 --cache-idle-slots --fit off --samplers top_k;top_p;min_p;temperature \
--ctx-size 200000 --model ./models/Qwen3.6-35B-A3B-Q6_K-00001-of-00002.gguf -ngl 99 \
--min-p 0.00 --top-p 0.95 --top-k 20 --chat-template-kwargs '{"preserve_thinking": "true"}' \
--tensor-split 2,1 --main-gpu 0 \
--seed 42 --temp 0.0 --cache-reuse 256
```

Is the hybrid architecture of Qwen3.6-35B-A3B the reason or I'm doing something wrong?

It feels that I'm expecting exactly what the [card](https://ggml.ai/f0.png) for the option describes, meanwhile I found a [comment](https://github.com/ggml-org/llama.cpp/pull/19408#issuecomment-3866421943) elsewhere that IMHO seems to be in direct contradiction with the card.

I also tried with a different number `--cache-reuse 2048` which didn't change anything. Here's a log from llama-server when a plan is prepared and output, just before I switch to Build agent and ask for implementation:

EDIT: I also tried with `--swa-full` but that didn't help either.

<details>

```
slot init_sampler: id  1 | task 1305 | init sampler, took 2.55 ms, tokens: text = 29078, total = 29078
slot update_slots: id  1 | task 1305 | prompt processing done, n_tokens = 29078, batch.n_tokens = 4
slot create_check: id  1 | task 1305 | created context checkpoint 12 of 128 (pos_min = 29073, pos_max = 29073, n_tokens = 29074, size = 62.813 MiB)
srv  log_server_r: done request: POST /v1/chat/completions 172.17.0.2 200
slot print_timing: id  1 | task 1305 | 
prompt eval time =     934.50 ms /   575 tokens (    1.63 ms per token,   615.30 tokens per second)
       eval time =   24405.81 ms /  1078 tokens (   22.64 ms per token,    44.17 tokens per second)
      total time =   25340.31 ms /  1653 tokens
slot      release: id  1 | task 1305 | stop processing: n_tokens = 30155, truncated = 0
srv  update_slots: all slots are idle

###
# switch agent to Build + "please implement"
###

srv  params_from_: Chat format: peg-native
slot get_availabl: id  1 | task -1 | selected slot by LCP similarity, sim_best = 0.187 (> 0.100 thold), f_keep = 0.186
srv  get_availabl: updating prompt cache
srv   prompt_save:  - saving prompt with length 30155, total state size = 652.353 MiB
srv          load:  - looking for better prompt, base f_keep = 0.186, sim = 0.187
srv        update:  - cache state: 1 prompts, 1406.112 MiB (limits: 16384.000 MiB, 200192 tokens, 351365 est)
srv        update:    - prompt 0x34ea97e0:   30155 tokens, checkpoints: 12,  1406.112 MiB
srv  get_availabl: prompt cache update took 842.32 ms
slot launch_slot_: id  1 | task -1 | sampler chain: logits -> top-k -> ?top-p -> ?min-p -> temp-ext -> dist 
slot launch_slot_: id  1 | task 2386 | processing task, is_child = 0
slot update_slots: id  1 | task 2386 | new prompt, n_ctx_slot = 200192, n_keep = 0, task.n_tokens = 29956
slot update_slots: id  1 | task 2386 | n_past = 5610, slot.prompt.tokens.size() = 30155, seq_id = 1, pos_min = 30154, n_swa = 0
slot update_slots: id  1 | task 2386 | Checking checkpoint with [29073, 29073] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [28561, 28561] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [28294, 28294] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [27782, 27782] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [25818, 25818] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [25306, 25306] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [22964, 22964] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [22452, 22452] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [14569, 14569] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [14057, 14057] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [13566, 13566] against 5610...
slot update_slots: id  1 | task 2386 | Checking checkpoint with [13054, 13054] against 5610...
slot update_slots: id  1 | task 2386 | forcing full prompt re-processing due to lack of cache data (likely due to SWA or hybrid/recurrent memory, see https://github.com/ggml-org/llama.cpp/pull/13194#issuecomment-2868343055)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 13054, pos_max = 13054, n_tokens = 13055, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 13566, pos_max = 13566, n_tokens = 13567, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 14057, pos_max = 14057, n_tokens = 14058, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 14569, pos_max = 14569, n_tokens = 14570, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 22452, pos_max = 22452, n_tokens = 22453, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 22964, pos_max = 22964, n_tokens = 22965, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 25306, pos_max = 25306, n_tokens = 25307, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 25818, pos_max = 25818, n_tokens = 25819, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 27782, pos_max = 27782, n_tokens = 27783, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 28294, pos_max = 28294, n_tokens = 28295, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 28561, pos_max = 28561, n_tokens = 28562, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | erased invalidated context checkpoint (pos_min = 29073, pos_max = 29073, n_tokens = 29074, n_swa = 0, pos_next = 0, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | n_tokens = 0, memory_seq_rm [0, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 2048, batch.n_tokens = 2048, progress = 0.068367
slot update_slots: id  1 | task 2386 | n_tokens = 2048, memory_seq_rm [2048, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 4096, batch.n_tokens = 2048, progress = 0.136734
slot update_slots: id  1 | task 2386 | n_tokens = 4096, memory_seq_rm [4096, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 6144, batch.n_tokens = 2048, progress = 0.205101
slot update_slots: id  1 | task 2386 | n_tokens = 6144, memory_seq_rm [6144, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 8192, batch.n_tokens = 2048, progress = 0.273468
slot update_slots: id  1 | task 2386 | n_tokens = 8192, memory_seq_rm [8192, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 10240, batch.n_tokens = 2048, progress = 0.341835
slot update_slots: id  1 | task 2386 | n_tokens = 10240, memory_seq_rm [10240, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 12288, batch.n_tokens = 2048, progress = 0.410202
slot update_slots: id  1 | task 2386 | n_tokens = 12288, memory_seq_rm [12288, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 14336, batch.n_tokens = 2048, progress = 0.478569
slot update_slots: id  1 | task 2386 | n_tokens = 14336, memory_seq_rm [14336, end)
slot update_slots: id  1 | task 2386 | 13000 tokens since last checkpoint at 0, creating new checkpoint during processing at position 16384
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 16384, batch.n_tokens = 2048, progress = 0.546935
slot create_check: id  1 | task 2386 | created context checkpoint 1 of 128 (pos_min = 14335, pos_max = 14335, n_tokens = 14336, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | n_tokens = 16384, memory_seq_rm [16384, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 18432, batch.n_tokens = 2048, progress = 0.615302
slot update_slots: id  1 | task 2386 | n_tokens = 18432, memory_seq_rm [18432, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 20480, batch.n_tokens = 2048, progress = 0.683669
slot update_slots: id  1 | task 2386 | n_tokens = 20480, memory_seq_rm [20480, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 22528, batch.n_tokens = 2048, progress = 0.752036
slot update_slots: id  1 | task 2386 | n_tokens = 22528, memory_seq_rm [22528, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 24576, batch.n_tokens = 2048, progress = 0.820403
slot update_slots: id  1 | task 2386 | n_tokens = 24576, memory_seq_rm [24576, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 26624, batch.n_tokens = 2048, progress = 0.888770
slot update_slots: id  1 | task 2386 | n_tokens = 26624, memory_seq_rm [26624, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 28672, batch.n_tokens = 2048, progress = 0.957137
slot update_slots: id  1 | task 2386 | n_tokens = 28672, memory_seq_rm [28672, end)
slot update_slots: id  1 | task 2386 | 13000 tokens since last checkpoint at 14336, creating new checkpoint during processing at position 29440
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 29440, batch.n_tokens = 768, progress = 0.982775
slot create_check: id  1 | task 2386 | created context checkpoint 2 of 128 (pos_min = 28671, pos_max = 28671, n_tokens = 28672, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | n_tokens = 29440, memory_seq_rm [29440, end)
slot update_slots: id  1 | task 2386 | prompt processing progress, n_tokens = 29952, batch.n_tokens = 512, progress = 0.999866
slot create_check: id  1 | task 2386 | created context checkpoint 3 of 128 (pos_min = 29439, pos_max = 29439, n_tokens = 29440, size = 62.813 MiB)
slot update_slots: id  1 | task 2386 | n_tokens = 29952, memory_seq_rm [29952, end)
reasoning-budget: activated, budget=2147483647 tokens
reasoning-budget: deactivated (natural end)
slot init_sampler: id  1 | task 2386 | init sampler, took 2.62 ms, tokens: text = 29956, total = 29956
slot update_slots: id  1 | task 2386 | prompt processing done, n_tokens = 29956, batch.n_tokens = 4
slot create_check: id  1 | task 2386 | created context checkpoint 4 of 128 (pos_min = 29951, pos_max = 29951, n_tokens = 29952, size = 62.813 MiB)
srv  log_server_r: done request: POST /v1/chat/completions 172.17.0.2 200
slot print_timing: id  1 | task 2386 | 
prompt eval time =   16817.75 ms / 29956 tokens (    0.56 ms per token,  1781.21 tokens per second)
       eval time =    6227.23 ms /   281 tokens (   22.16 ms per token,    45.12 tokens per second)
      total time =   23044.98 ms / 30237 tokens
slot      release: id  1 | task 2386 | stop processing: n_tokens = 30236, truncated = 0
srv  update_slots: all slots are idle
```

</details>

**Comment by Farmadupe** [comment, marked as answer]
FYI `--cache-reuse` is misleadingly named. It's a processing trick to allow you to keep the tail end of prefilled sequence after deleting earlier parts, but crucially while  **pretending that the earlier parts are still there**. Since you probably don't want any trace of the old system prompt to be hanging around `--cache-reuse` is almost definitely not what you want.

The linked comment mentions that `--cache-reuse` doesn't work with models that use recurrent state. `Qwen3.6-35B-A3B` does use recurrent state, so your findings actually are consistent with what `ggerganov` wrote.

`Qwen3.6-35B-A3B` does not using sliding window attention, so `--swa-full` has no effect. 

To answer your original question though, if you want to replace your original system prompt and have it take effect, then unfortunately the only way is to reprocess the entire history.

**Reply by Nindaleth** [reply]
Dang, doing another prefill pass on a 100k context window sucks on my env. Fortunately with OpenCode I can adapt the flow to work around this (either use obra/superpowers and start right away with the Build agent, or use a subagent for the read-only part of the work)

Thanks for the explainer!

**Comment by fillumina** [comment]
I ran into the same issue and found a solution that requires no proxy or code changes.

It turns out `id_slot` passes through llama-server's `/v1/chat/completions` endpoint even though it's undocumented there. You can verify it with a quick curl — set `id_slot: 1` in the body and check the server log, the request lands on that slot.

The second piece is that opencode forwards the `options` block of a model config directly into the request body. So you can define two model entries pointing to the same GGUF file with different `id_slot` values, and route all built-in subagents (`general`, `explorimage of a queue with two serverse`, `summary`, `compaction`, `title`) to the second one via the `agent` config.

The main session stays on slot 0, all subagents share slot 1. Neither ever evicts the other's KV cache.

I wrote it up in detail here: https://fillumina.wordpress.com/2026/05/12/optimize-opencode-for-better-parallelism-and-efficiency-with-llama-server/

**Reply by Nindaleth** [reply]
Hey, thanks for looking into this issue, I've read your article.

Unfortunately, your solution does not apply to the problem that I have - you can see in my report:
* processing never changed slots, everything took place in slot 1 and yet still I had to reprocess everything
* I never used a subagent, everything ran in the main agent chat, the switch from Plan to Build was just one Tab press in OpenCode

Hence applying slot magic to work around subagents calls _is_ pretty cool, but not useful to my issue.

Couple more comments to your article:

> When a new request arrives the server picks a slot using Longest Common Prefix (LCP) matching

That's what the selected answer points out; by switching from Plan to Build I changed the initial system prompt which modified the Prefix and invalidated all caches, that's why I need to reprocess the entire context.

> They are also serialized: opencode doesn’t fire two subagents in parallel.

This can't be true, otherwise I wouldn't be able to spend about half my March's Copilot premium request allowance launching 8 subagents in parallel in OpenCode, each taking 15-20 minutes to finish its work, while the full session took slightly over 20 minutes in total.

**Reply by fillumina** [reply]
Hi,
opencode calls subagents (general, explore, summary, compaction and title) when it performs normal tasks in the main agent even if it doesn't notify it.  For example in Plan mode if it needs to explore the project structure it calls the explore agent which will execute in its own context, do its research and return the results to the main agent. By default both agents will refer to the same context on the model (because of a limitation in the openai protocol that doesn't allow to define a different context for each agent, except that the llama-server implementation allows for exactly that). These subagent callings count as context switching and are the cause of the context reloading you are observing. My technique, through the use of parallel sessions reserves the caller context intact in case of a subagent call.
For what concern the fact that agents are executed in parallel, thy are if configured so, but subagents called from an agent aren't (I am referring to the standard ones: general, explore, summary, compaction and title).
At least this is my understanding of things right now.

**Reply by Nindaleth** [reply]
My OpenCode always notifies me when running a subagent task (because it's a Task tool usage and OpenCode reports all tool usage), there's nothing hidden happening behind the scenes in OpenCode in this regard. My issue happens even if I completely forbid subagents (via denying Task tool usage) because it's not _sub_ agent dependent. Feel free to check the `details` section in my original post where I don't use any subagents and all serial requests from a single agent land into the same slot.

I'm not disputing the merit of your solution BTW, it solves a real problem.

**Reply by fillumina** [reply]
You are absolutely right, I am sorry, I should have read your post more carefully. These are different problems because your issue is caused by the agent prompt, which changes, mine by a context reloading after using other agents.  Btw I checked and the main agent (plan or build) doesn't invoke the explore agent all the times but only if it feels there are a lot of documents to explore, and yes, it shows in the chat.

Actually, following your lead, I am exploring that --swa-full option too to see if it might be beneficial, context reloading is a huge source of time waste for these small local models. 

Hope to hear some good news from your experiments, sorry again for the confusion.
